File: backend/app/schemas/wealth.py
"""Wealth API schemas — budget, expenses, vaults, investments, dashboard."""
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel
from typing import Optional


# ----- Budget -----
class BudgetCategoryResponse(BaseModel):
    id: UUID
    key: str
    label: str
    target_percentage: float
    display_order: Optional[int] = None

    class Config:
        from_attributes = True


class BudgetAllocationCreate(BaseModel):
    category_key: str
    target_percentage: float


class BudgetAllocationResponse(BaseModel):
    id: UUID
    finance_profile_id: UUID
    category_key: str
    target_percentage: float

    class Config:
        from_attributes = True


# ----- Expenses -----
class ExpenseCreate(BaseModel):
    amount: float
    currency: Optional[str] = "USD"
    category: str  # needs, wants, savings
    description: Optional[str] = None


class ExpenseResponse(BaseModel):
    id: UUID
    finance_profile_id: UUID
    amount: float
    currency: Optional[str]
    category: str
    description: Optional[str]
    occurred_at: datetime

    class Config:
        from_attributes = True


# ----- Income -----
class IncomeEntryCreate(BaseModel):
    amount: float
    currency: Optional[str] = "USD"
    income_source_id: Optional[UUID] = None
    notes: Optional[str] = None


class IncomeEntryResponse(BaseModel):
    id: UUID
    finance_profile_id: UUID
    amount: float
    currency: Optional[str]
    income_source_id: Optional[UUID]
    occurred_at: datetime
    automation_applied: Optional[int]

    class Config:
        from_attributes = True


# ----- Vaults -----
class MoneyVaultCreate(BaseModel):
    name: str
    description: Optional[str] = None
    target_amount: float
    currency: Optional[str] = "USD"
    unlock_date: datetime
    allocation_key: Optional[str] = None


class MoneyVaultUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    target_amount: Optional[float] = None
    unlock_date: Optional[datetime] = None


class MoneyVaultAddRequest(BaseModel):
    amount: float


class MoneyVaultResponse(BaseModel):
    id: UUID
    finance_profile_id: UUID
    name: str
    description: Optional[str]
    target_amount: float
    current_amount: float
    currency: Optional[str]
    unlock_date: datetime
    allocation_key: Optional[str]

    class Config:
        from_attributes = True


# ----- Investment -----
class InvestmentAllocationCreate(BaseModel):
    allocation_key: str
    label: Optional[str] = None
    percentage: float


class InvestmentStrategyCreate(BaseModel):
    name: str
    label: Optional[str] = None
    allocations: list[InvestmentAllocationCreate]


class InvestmentStrategyResponse(BaseModel):
    id: UUID
    finance_profile_id: UUID
    name: str
    label: Optional[str]
    is_active: Optional[int]
    allocations: list[dict]

    class Config:
        from_attributes = True


class InvestmentAccountCreate(BaseModel):
    name: str
    account_type: Optional[str] = None
    allocation_key: Optional[str] = None
    currency: Optional[str] = "USD"


class InvestmentAccountResponse(BaseModel):
    id: UUID
    finance_profile_id: UUID
    name: str
    account_type: Optional[str]
    allocation_key: Optional[str]
    balance: float
    currency: Optional[str]

    class Config:
        from_attributes = True


# ----- Dashboard -----
class BudgetDistributionItem(BaseModel):
    category_key: str
    label: str
    target_percentage: float
    spent_amount: float
    spent_percentage: float
    is_over: bool


class WealthDashboardResponse(BaseModel):
    net_worth: Optional[float] = None
    total_income_this_month: Optional[float] = None
    budget_distribution: list[BudgetDistributionItem] = []
    vaults: list[MoneyVaultResponse] = []
    investment_accounts: list[InvestmentAccountResponse] = []
    active_strategy: Optional[InvestmentStrategyResponse] = None

File: frontend/src/services/wealth.ts
import { api } from './api'

export interface BudgetCategory {
  id: string
  key: string
  label: string
  target_percentage: number
  display_order?: number
}

export interface BudgetAllocation {
  id: string
  finance_profile_id: string
  category_key: string
  target_percentage: number
}

export interface Expense {
  id: string
  finance_profile_id: string
  amount: number
  currency?: string
  category: string
  description?: string
  occurred_at: string
}

export interface IncomeEntry {
  id: string
  finance_profile_id: string
  amount: number
  currency?: string
  income_source_id?: string
  occurred_at: string
  automation_applied?: number
}

export interface MoneyVault {
  id: string
  finance_profile_id: string
  name: string
  description?: string
  target_amount: number
  current_amount: number
  currency?: string
  unlock_date: string
  allocation_key?: string
}

export interface InvestmentAllocationItem {
  allocation_key: string
  label?: string
  percentage: number
}

export interface InvestmentStrategy {
  id: string
  finance_profile_id: string
  name: string
  label?: string
  is_active?: number
  allocations: InvestmentAllocationItem[]
}

export interface InvestmentAccount {
  id: string
  finance_profile_id: string
  name: string
  account_type?: string
  allocation_key?: string
  balance: number
  currency?: string
}

export interface BudgetDistributionItem {
  category_key: string
  label: string
  target_percentage: number
  spent_amount: number
  spent_percentage: number
  is_over: boolean
}

export interface WealthDashboard {
  net_worth?: number
  total_income_this_month?: number
  budget_distribution: BudgetDistributionItem[]
  vaults: MoneyVault[]
  investment_accounts: InvestmentAccount[]
  active_strategy?: InvestmentStrategy | null
}

export const wealthService = {
  getDashboard: () => api.get<WealthDashboard>('/wealth/dashboard'),
  listBudgetCategories: () => api.get<BudgetCategory[]>('/wealth/budget-categories'),
  listBudgetAllocations: () => api.get<BudgetAllocation[]>('/wealth/budget-allocations'),
  upsertBudgetAllocation: (data: { category_key: string; target_percentage: number }) =>
    api.post<BudgetAllocation>('/wealth/budget-allocations', data),
  listExpenses: (params?: { category?: string; month?: string }) => {
    const q = new URLSearchParams()
    if (params?.category) q.set('category', params.category)
    if (params?.month) q.set('month', params.month)
    const query = q.toString()
    return api.get<Expense[]>(`/wealth/expenses${query ? `?${query}` : ''}`)
  },
  createExpense: (data: { amount: number; currency?: string; category: string; description?: string }) =>
    api.post<Expense>('/wealth/expenses', data),
  listIncomeEntries: () => api.get<IncomeEntry[]>('/wealth/income-entries'),
  createIncomeEntry: (data: { amount: number; currency?: string; income_source_id?: string; notes?: string }) =>
    api.post<IncomeEntry>('/wealth/income-entries', data),
  listIncomeSources: () => api.get<{ id: string; name: string; amount_monthly?: number; currency?: string }[]>('/wealth/income-sources'),
  listVaults: () => api.get<MoneyVault[]>('/wealth/savings-vaults'),
  createVault: (data: {
    name: string
    description?: string
    target_amount: number
    currency?: string
    unlock_date: string
    allocation_key?: string
  }) => api.post<MoneyVault>('/wealth/savings-vaults', data),
  addToVault: (vaultId: string, amount: number) =>
    api.post<MoneyVault>(`/wealth/savings-vaults/${vaultId}/add`, { amount }),
  updateVault: (vaultId: string, data: Partial<{ name: string; description: string; target_amount: number; unlock_date: string }>) =>
    api.patch<MoneyVault>(`/wealth/savings-vaults/${vaultId}`, data),
  listStrategies: () => api.get<InvestmentStrategy[]>('/wealth/strategies'),
  createStrategy: (data: {
    name: string
    label?: string
    allocations: { allocation_key: string; label?: string; percentage: number }[]
  }) => api.post<InvestmentStrategy>('/wealth/strategies', data),
  activateStrategy: (strategyId: string) =>
    api.post<InvestmentStrategy>(`/wealth/strategies/${strategyId}/activate`, {}),
  listInvestmentAccounts: () => api.get<InvestmentAccount[]>('/wealth/investment-accounts'),
  createInvestmentAccount: (data: { name: string; account_type?: string; allocation_key?: string; currency?: string }) =>
    api.post<InvestmentAccount>('/wealth/investment-accounts', data),
}
